File: sdk-py/noether_sdk/sub/vaults.py
from ..errors import NotFoundError
from ..models import VaultActivityRow, VaultRow
from ..transport import Transport


class VaultsApi:
    def __init__(self, transport: Transport) -> None:
        self._transport = transport

    async def list(
        self, *, leader: str | None = None, limit: int | None = None
    ) -> list[VaultRow]:
        body = await self._transport.request(
            "GET", "/v1/vaults", params={"leader": leader, "limit": limit}
        )
        return [VaultRow.model_validate(v) for v in body.get("vaults", [])]

    async def get(self, vault_id: int) -> VaultRow | None:
        try:
            body = await self._transport.request("GET", f"/v1/vaults/{vault_id}")
        except NotFoundError:
            return None
        return VaultRow.model_validate(body)

    async def deposits(self, vault_id: int, *, limit: int | None = None) -> list[VaultActivityRow]:
        body = await self._transport.request(
            "GET", f"/v1/vaults/{vault_id}/deposits", params={"limit": limit}
        )
        return [VaultActivityRow.model_validate(r) for r in body.get("deposits", [])]

    async def withdraws(self, vault_id: int, *, limit: int | None = None) -> list[VaultActivityRow]:
        body = await self._transport.request(
            "GET", f"/v1/vaults/{vault_id}/withdraws", params={"limit": limit}
        )
        return [VaultActivityRow.model_validate(r) for r in body.get("withdraws", [])]

    async def fee_claims(
        self, vault_id: int, *, limit: int | None = None
    ) -> list[VaultActivityRow]:
        body = await self._transport.request(
            "GET", f"/v1/vaults/{vault_id}/fee-claims", params={"limit": limit}
        )
        return [VaultActivityRow.model_validate(r) for r in body.get("feeClaims", [])]

File: sdk-ts/src/sub/vaults.ts
import type { Transport } from '../transport.js';

export interface VaultRow {
  id: number;
  leader: string;
  name: string;
  createdAt: number;
  totalUsdc: string;
  circulatingShares: string;
  hwmNav: string;
  realizedPnl: string;
  leaderShares: string;
  profitShareBps: number;
  paused: boolean;
  updatedAt: number;
}

export interface VaultActivityRow {
  id: number;
  vaultId: number;
  principal: string;
  amount: string;
  shares?: string;
  ledger: number;
  ts: number;
  txHash: string;
}

export interface VaultTradeRow {
  id: number;
  vaultId: number;
  positionId: string;
  action: 'open' | 'close';
  leader: string;
  collateral: string;
  /** Settled PnL on close rows (7-dec USDC, signed); null on open rows. */
  pnl?: string | null;
  ledger: number;
  ts: number;
  txHash: string;
}

export interface VaultListQuery {
  leader?: string;
  limit?: number;
}

export interface VaultActivityQuery {
  limit?: number;
}

export class VaultsApi {
  constructor(private readonly transport: Transport) {}

  async list(query: VaultListQuery = {}): Promise<VaultRow[]> {
    const res = await this.transport.request<{ vaults: VaultRow[] }>({
      path: '/v1/vaults',
      query: { leader: query.leader, limit: query.limit },
    });
    return res.vaults;
  }

  async get(id: number): Promise<VaultRow> {
    return this.transport.request<VaultRow>({ path: `/v1/vaults/${id}` });
  }

  async trades(id: number, query: VaultActivityQuery = {}): Promise<VaultTradeRow[]> {
    const res = await this.transport.request<{ trades: VaultTradeRow[] }>({
      path: `/v1/vaults/${id}/trades`,
      query: { limit: query.limit },
    });
    return res.trades;
  }

  async deposits(id: number, query: VaultActivityQuery = {}): Promise<VaultActivityRow[]> {
    const res = await this.transport.request<{ deposits: VaultActivityRow[] }>({
      path: `/v1/vaults/${id}/deposits`,
      query: { limit: query.limit },
    });
    return res.deposits;
  }

  async withdraws(id: number, query: VaultActivityQuery = {}): Promise<VaultActivityRow[]> {
    const res = await this.transport.request<{ withdraws: VaultActivityRow[] }>({
      path: `/v1/vaults/${id}/withdraws`,
      query: { limit: query.limit },
    });
    return res.withdraws;
  }

  async feeClaims(id: number, query: VaultActivityQuery = {}): Promise<VaultActivityRow[]> {
    const res = await this.transport.request<{ feeClaims: VaultActivityRow[] }>({
      path: `/v1/vaults/${id}/fee-claims`,
      query: { limit: query.limit },
    });
    return res.feeClaims;
  }
}
